storjscandb/headers.go

```diff
@@ -0,0 +1,115 @@
+	"database/sql"
+	"time"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storjscan/blockchain"
+	"storj.io/storjscan/storjscandb/dbx"
+)
+
+// ErrHeadersDB indicates about internal headers DB error.
+var ErrHeadersDB = errs.Class("HeadersDB")
+
+// ensures that headersDB implements blockchain.HeadersDB.
+var _ blockchain.HeadersDB = (*headersDB)(nil)
+
+// headersDB is headers database cache dbx postgres implementation.
+//
+// architecture: Database
+type headersDB struct {
+	db *dbx.DB
+}
+
+// Insert inserts new block header into db.
+func (headers *headersDB) Insert(ctx context.Context, hash blockchain.Hash, number int64, timestamp time.Time) error {
+	_, err := headers.db.Create_BlockHeader(ctx,
+		dbx.BlockHeader_Hash(hash.Bytes()),
+		dbx.BlockHeader_Number(number),
+		dbx.BlockHeader_Timestamp(timestamp.UTC()))
+
+	return ErrHeadersDB.Wrap(err)
+}
+
+// Delete deletes block header from the db by block hash.
+func (headers *headersDB) Delete(ctx context.Context, hash blockchain.Hash) error {
+	deleted, err := headers.db.Delete_BlockHeader_By_Hash(ctx, dbx.BlockHeader_Hash(hash.Bytes()))
+	if err != nil {
+		return ErrHeadersDB.Wrap(err)
+	}
+	if !deleted {
+		return blockchain.ErrNoHeader
+	}
+	return nil
+}
+
+// Get retrieves single block header from the db by block hash.
+func (headers *headersDB) Get(ctx context.Context, hash blockchain.Hash) (blockchain.Header, error) {
+	dbxHeader, err := headers.db.Get_BlockHeader_By_Hash(ctx, dbx.BlockHeader_Hash(hash.Bytes()))
+	if err != nil {
+		if errs.Is(err, sql.ErrNoRows) {
+			return blockchain.Header{}, blockchain.ErrNoHeader
+		}
+
+		return blockchain.Header{}, ErrHeadersDB.Wrap(err)
+	}
+
+	return fromDBXHeader(dbxHeader), nil
+}
+
+// GetByNumber retrieves single block header from the db by block number.
+func (headers *headersDB) GetByNumber(ctx context.Context, number int64) (blockchain.Header, error) {
+	dbxHeader, err := headers.db.Get_BlockHeader_By_Number(ctx, dbx.BlockHeader_Number(number))
+	if err != nil {
+		if errs.Is(err, sql.ErrNoRows) {
+			return blockchain.Header{}, blockchain.ErrNoHeader
+		}
+
+		return blockchain.Header{}, ErrHeadersDB.Wrap(err)
+	}
+
+	return fromDBXHeader(dbxHeader), nil
+}
+
+// After retrieves block header which block timestamp is after provided time.
+func (headers *headersDB) After(ctx context.Context, t time.Time) (blockchain.Header, error) {
+	dbxHeader, err := headers.db.First_BlockHeader_By_Timestamp_Greater(ctx, dbx.BlockHeader_Timestamp(t))
+	if err != nil {
+		return blockchain.Header{}, ErrHeadersDB.Wrap(err)
+	}
+	if dbxHeader == nil {
+		return blockchain.Header{}, blockchain.ErrNoHeader
+	}
+
+	return fromDBXHeader(dbxHeader), nil
+}
+
+// List retrieves all block headers from the db.
+func (headers *headersDB) List(ctx context.Context) ([]blockchain.Header, error) {
+	dbxHeaders, err := headers.db.All_BlockHeader_OrderBy_Desc_Timestamp(ctx)
+	if err != nil {
+		return nil, ErrHeadersDB.Wrap(err)
+	}
+
+	var list []blockchain.Header
```

Comment:
should we pre size the list here?

Reply:
I think list length is unknown here, and making a db call to get the count will be less efficient for small list. 

Right now List method is just for convenience, not sure if production code uses it somewhere. If it will this method will be updated to use limit and offset for pagination.

And even if the length known in advance and if preallocation doesn't give significant performance gain I think preallocation can be omitted for the sake of simplicity

Reply:
The size is known after the db call, so instead of 
`var list []blockchan.Header`
you could use 
`list := make([]blockchain.Header, 0 , len(dbxHeaders))`
This will pre size the list value so that it doesn't need to perform unnecessary memory allocations.